File: ports/go/lombokfuzzer.go
// Package lombokfuzzer is a Go port of LombokFuzzer — universal fuzzing framework.
//
// Part of the Lombok Ecosystem (github.com/codinglombok/LombokFuzzer).
// Apache-2.0 License.
package lombokfuzzer

import (
	"crypto/rand"
	"encoding/binary"
	"fmt"
	"hash/fnv"
	"math"
	"sync"
	"sync/atomic"
	"time"
)

// ─── Types ──────────────────────────────────────────────────────────────────

// FuzzMode selects the fuzzing strategy.
type FuzzMode string

const (
	ModeMutation     FuzzMode = "mutation"
	ModeGeneration   FuzzMode = "generation"
	ModeHybrid       FuzzMode = "hybrid"
	ModeDifferential FuzzMode = "differential"
	ModeDirected     FuzzMode = "directed"
)

// Severity classifies crash severity.
type Severity string

const (
	SeverityCritical Severity = "critical"
	SeverityHigh     Severity = "high"
	SeverityMedium   Severity = "medium"
	SeverityLow      Severity = "low"
	SeverityInfo     Severity = "info"
)

// FuzzConfig configures a fuzzing campaign.
type FuzzConfig struct {
	Name            string
	Mode            FuzzMode
	Seed            uint64
	MaxExecutions   uint64
	MaxTimeSeconds  int
	MaxInputSize    int
	TimeoutMs       int
	Workers         int
	TargetFunc      func(data []byte) error
}

// FuzzInput represents one test input.
type FuzzInput struct {
	Data       []byte
	Hash       uint64
	Depth      int
	Energy     float64
	Executions uint64
	CreatedAt  time.Time
}

// ExecutionResult captures what happened when a target processed an input.
type ExecutionResult struct {
	Input      *FuzzInput
	Crashed    bool
	NewCoverage bool
	DurationUs int64
	Error      error
}

// FuzzStats tracks campaign statistics.
type FuzzStats struct {
	TotalExecutions  uint64
	ExecsPerSecond   float64
	CorpusSize       int
	UniqueCrashes    int
	EdgesFound       int
	EdgesTotal       int
	CoveragePercent  float64
	ElapsedMs        int64
}

// ─── PRNG (xoshiro256**) ────────────────────────────────────────────────────

// PRNG implements xoshiro256** for fast, reproducible pseudo-random numbers.
type PRNG struct {
	s [4]uint64
}

// NewPRNG creates a seeded PRNG.
func NewPRNG(seed uint64) *PRNG {
	p := &PRNG{}
	// SplitMix64 seeding
	z := seed
	for i := 0; i < 4; i++ {
		z += 0x9e3779b97f4a7c15
		z = (z ^ (z >> 30)) * 0xbf58476d1ce4e5b9
		z = (z ^ (z >> 27)) * 0x94d049bb133111eb
		p.s[i] = z ^ (z >> 31)
	}
	return p
}

func (p *PRNG) rotl(x uint64, k uint) uint64 {
	return (x << k) | (x >> (64 - k))
}

// Next returns the next pseudo-random uint64.
func (p *PRNG) Next() uint64 {
	result := p.rotl(p.s[1]*5, 7) * 9
	t := p.s[1] << 17
	p.s[2] ^= p.s[0]
	p.s[3] ^= p.s[1]
	p.s[1] ^= p.s[2]
	p.s[0] ^= p.s[3]
	p.s[2] ^= t
	p.s[3] = p.rotl(p.s[3], 45)
	return result
}

// NextRange returns a random number in [0, n).
func (p *PRNG) NextRange(n int) int {
	if n <= 0 {
		return 0
	}
	return int(p.Next() % uint64(n))
}

// RandomBytes returns n random bytes.
func (p *PRNG) RandomBytes(n int) []byte {
	buf := make([]byte, n)
	for i := range buf {
		buf[i] = byte(p.Next())
	}
	return buf
}

// ─── Hash (FNV-1a 64-bit) ──────────────────────────────────────────────────

// FNV1a64 computes the FNV-1a 64-bit hash of data.
func FNV1a64(data []byte) uint64 {
	h := fnv.New64a()
	h.Write(data)
	return h.Sum64()
}

// ─── Coverage Tracker ───────────────────────────────────────────────────────

// CoverageTracker tracks edge coverage using a fixed-size bitmap.
type CoverageTracker struct {
	bitmap     []byte
	bitmapSize int
	edgesFound int
	mu         sync.RWMutex
}

// NewCoverageTracker creates a tracker with the given bitmap size.
func NewCoverageTracker(size int) *CoverageTracker {
	if size <= 0 {
		size = 65536
	}
	return &CoverageTracker{
		bitmap:     make([]byte, size),
		bitmapSize: size,
	}
}

// RecordEdge records a hit on an edge and returns true if it's new.
func (ct *CoverageTracker) RecordEdge(from, to uint64) bool {
	idx := int((from ^ to) % uint64(ct.bitmapSize))
	ct.mu.Lock()
	defer ct.mu.Unlock()
	if ct.bitmap[idx] == 0 {
		ct.bitmap[idx] = 1
		ct.edgesFound++
		return true
	}
	ct.bitmap[idx]++
	return false
}

// EdgesFound returns the number of unique edges discovered.
func (ct *CoverageTracker) EdgesFound() int {
	ct.mu.RLock()
	defer ct.mu.RUnlock()
	return ct.edgesFound
}

// CoveragePercent returns % of bitmap occupied.
func (ct *CoverageTracker) CoveragePercent() float64 {
	ct.mu.RLock()
	defer ct.mu.RUnlock()
	return float64(ct.edgesFound) / float64(ct.bitmapSize) * 100
}

// Reset clears the bitmap.
func (ct *CoverageTracker) Reset() {
	ct.mu.Lock()
	defer ct.mu.Unlock()
	for i := range ct.bitmap {
		ct.bitmap[i] = 0
	}
	ct.edgesFound = 0
}

// ─── Mutator Engine ─────────────────────────────────────────────────────────

// MutatorStrategy names a mutation strategy.
type MutatorStrategy string

const (
	StrategyBitFlip    MutatorStrategy = "bit_flip"
	StrategyByteFlip   MutatorStrategy = "byte_flip"
	StrategyArithmetic MutatorStrategy = "arithmetic"
	StrategyInteresting MutatorStrategy = "interesting"
	StrategyHavoc      MutatorStrategy = "havoc"
	StrategySplice     MutatorStrategy = "splice"
)

// MutatorEngine applies mutations to inputs.
type MutatorEngine struct {
	prng       *PRNG
	strategies []MutatorStrategy
}

// NewMutatorEngine creates a mutator with the given PRNG.
func NewMutatorEngine(prng *PRNG) *MutatorEngine {
	return &MutatorEngine{
		prng: prng,
		strategies: []MutatorStrategy{
			StrategyBitFlip, StrategyByteFlip, StrategyArithmetic,
			StrategyInteresting, StrategyHavoc,
		},
	}
}

// Mutate applies a random mutation to the input and returns a new buffer.
func (me *MutatorEngine) Mutate(data []byte) []byte {
	if len(data) == 0 {
		return me.prng.RandomBytes(me.prng.NextRange(64) + 1)
	}

	out := make([]byte, len(data))
	copy(out, data)

	strategy := me.strategies[me.prng.NextRange(len(me.strategies))]
	switch strategy {
	case StrategyBitFlip:
		pos := me.prng.NextRange(len(out))
		bit := me.prng.NextRange(8)
		out[pos] ^= 1 << uint(bit)

	case StrategyByteFlip:
		pos := me.prng.NextRange(len(out))
		out[pos] ^= 0xFF

	case StrategyArithmetic:
		pos := me.prng.NextRange(len(out))
		delta := byte(me.prng.NextRange(70) - 35)
		out[pos] += delta

	case StrategyInteresting:
		interesting := []byte{0, 1, 0x7F, 0x80, 0xFF}
		pos := me.prng.NextRange(len(out))
		out[pos] = interesting[me.prng.NextRange(len(interesting))]

	case StrategyHavoc:
		rounds := me.prng.NextRange(8) + 1
		for r := 0; r < rounds; r++ {
			pos := me.prng.NextRange(len(out))
			out[pos] = byte(me.prng.Next())
		}
	}

	return out
}

// ─── Corpus Store ───────────────────────────────────────────────────────────

// CorpusStore manages the input corpus.
type CorpusStore struct {
	entries []*FuzzInput
	hashes  map[uint64]bool
	mu      sync.RWMutex
}

// NewCorpusStore creates an empty corpus.
func NewCorpusStore() *CorpusStore {
	return &CorpusStore{
		hashes: make(map[uint64]bool),
	}
}

// Add adds an input if it's not a duplicate. Returns true if added.
func (cs *CorpusStore) Add(input *FuzzInput) bool {
	cs.mu.Lock()
	defer cs.mu.Unlock()
	if cs.hashes[input.Hash] {
		return false
	}
	cs.hashes[input.Hash] = true
	cs.entries = append(cs.entries, input)
	return true
}

// Pick returns a random input from the corpus.
func (cs *CorpusStore) Pick(prng *PRNG) *FuzzInput {
	cs.mu.RLock()
	defer cs.mu.RUnlock()
	if len(cs.entries) == 0 {
		return nil
	}
	return cs.entries[prng.NextRange(len(cs.entries))]
}

// Size returns the corpus size.
func (cs *CorpusStore) Size() int {
	cs.mu.RLock()
	defer cs.mu.RUnlock()
	return len(cs.entries)
}

// ─── Crash Analyzer ─────────────────────────────────────────────────────────

// CrashInfo stores crash details.
type CrashInfo struct {
	ID           string
	Input        *FuzzInput
	Error        error
	Severity     Severity
	DiscoveredAt time.Time
	IsDuplicate  bool
}

// CrashAnalyzer detects and deduplicates crashes.
type CrashAnalyzer struct {
	crashes []CrashInfo
	hashes  map[uint64]bool
	mu      sync.Mutex
}

// NewCrashAnalyzer creates an empty analyzer.
func NewCrashAnalyzer() *CrashAnalyzer {
	return &CrashAnalyzer{
		hashes: make(map[uint64]bool),
	}
}

// Record records a crash. Returns true if it's unique.
func (ca *CrashAnalyzer) Record(input *FuzzInput, err error) bool {
	h := FNV1a64([]byte(err.Error()))
	ca.mu.Lock()
	defer ca.mu.Unlock()
	isDup := ca.hashes[h]
	ca.hashes[h] = true
	crash := CrashInfo{
		ID:           fmt.Sprintf("crash-%016x", h),
		Input:        input,
		Error:        err,
		Severity:     SeverityHigh,
		DiscoveredAt: time.Now(),
		IsDuplicate:  isDup,
	}
	ca.crashes = append(ca.crashes, crash)
	return !isDup
}

// UniqueCount returns the number of unique crashes.
func (ca *CrashAnalyzer) UniqueCount() int {
	ca.mu.Lock()
	defer ca.mu.Unlock()
	return len(ca.hashes)
}

// ─── Fuzz Engine ────────────────────────────────────────────────────────────

// FuzzEngine is the main fuzzing engine.
type FuzzEngine struct {
	config   FuzzConfig
	prng     *PRNG
	mutator  *MutatorEngine
	corpus   *CorpusStore
	coverage *CoverageTracker
	crashes  *CrashAnalyzer
	stats    FuzzStats
	stopCh   chan struct{}
	running  atomic.Bool
}

// New creates a new FuzzEngine.
func New(config FuzzConfig) *FuzzEngine {
	if config.MaxInputSize <= 0 {
		config.MaxInputSize = 4096
	}
	if config.Workers <= 0 {
		config.Workers = 1
	}
	prng := NewPRNG(config.Seed)
	return &FuzzEngine{
		config:   config,
		prng:     prng,
		mutator:  NewMutatorEngine(prng),
		corpus:   NewCorpusStore(),
		coverage: NewCoverageTracker(65536),
		crashes:  NewCrashAnalyzer(),
		stopCh:   make(chan struct{}),
	}
}

// Run starts the fuzzing campaign. Blocks until done or stopped.
func (fe *FuzzEngine) Run() FuzzStats {
	fe.running.Store(true)
	defer fe.running.Store(false)

	startTime := time.Now()

	// Seed corpus with one random input if empty
	if fe.corpus.Size() == 0 {
		seed := fe.prng.RandomBytes(fe.prng.NextRange(64) + 1)
		fe.corpus.Add(&FuzzInput{
			Data:      seed,
			Hash:      FNV1a64(seed),
			Energy:    1.0,
			CreatedAt: time.Now(),
		})
	}

	var execCount uint64

	for {
		select {
		case <-fe.stopCh:
			return fe.buildStats(startTime)
		default:
		}

		if fe.config.MaxExecutions > 0 && execCount >= fe.config.MaxExecutions {
			break
		}
		if fe.config.MaxTimeSeconds > 0 && time.Since(startTime) > time.Duration(fe.config.MaxTimeSeconds)*time.Second {
			break
		}

		// Pick & mutate
		parent := fe.corpus.Pick(fe.prng)
		if parent == nil {
			break
		}
		mutated := fe.mutator.Mutate(parent.Data)
		hash := FNV1a64(mutated)
		input := &FuzzInput{
			Data:      mutated,
			Hash:      hash,
			Depth:     parent.Depth + 1,
			Energy:    parent.Energy,
			CreatedAt: time.Now(),
		}

		// Execute target
		result := fe.execute(input)
		execCount++

		// Coverage feedback
		edgeHash := FNV1a64(mutated[:min(len(mutated), 8)])
		newEdge := fe.coverage.RecordEdge(uint64(parent.Hash), edgeHash)
		if newEdge {
			fe.corpus.Add(input)
		}

		// Crash detection
		if result.Crashed && result.Error != nil {
			fe.crashes.Record(input, result.Error)
		}
	}

	return fe.buildStats(startTime)
}

// Stop signals the engine to stop.
func (fe *FuzzEngine) Stop() {
	close(fe.stopCh)
}

func (fe *FuzzEngine) execute(input *FuzzInput) ExecutionResult {
	start := time.Now()
	var crashed bool
	var execErr error

	func() {
		defer func() {
			if r := recover(); r != nil {
				crashed = true
				execErr = fmt.Errorf("panic: %v", r)
			}
		}()
		if fe.config.TargetFunc != nil {
			if err := fe.config.TargetFunc(input.Data); err != nil {
				crashed = true
				execErr = err
			}
		}
	}()

	return ExecutionResult{
		Input:      input,
		Crashed:    crashed,
		DurationUs: time.Since(start).Microseconds(),
		Error:      execErr,
	}
}

func (fe *FuzzEngine) buildStats(startTime time.Time) FuzzStats {
	elapsed := time.Since(startTime).Milliseconds()
	var execsPerSec float64
	if elapsed > 0 {
		execsPerSec = math.Round(float64(fe.corpus.Size()) / float64(elapsed) * 1000)
	}
	return FuzzStats{
		TotalExecutions: fe.config.MaxExecutions,
		ExecsPerSecond:  execsPerSec,
		CorpusSize:      fe.corpus.Size(),
		UniqueCrashes:   fe.crashes.UniqueCount(),
		EdgesFound:      fe.coverage.EdgesFound(),
		EdgesTotal:      65536,
		CoveragePercent: fe.coverage.CoveragePercent(),
		ElapsedMs:       elapsed,
	}
}

// ─── Version ────────────────────────────────────────────────────────────────

const Version = "0.2.0"
const CodeName = "LombokFuzzer"

File: ports/python/lombokfuzzer.py
"""
LombokFuzzer — Python port of the universal fuzzing framework.

Part of the Lombok Ecosystem (github.com/codinglombok/LombokFuzzer).
Apache-2.0 License.
"""

from __future__ import annotations
import struct
import time
from dataclasses import dataclass, field
from typing import Callable, Optional

__version__ = "0.2.0"

# ─── PRNG (xoshiro256**) ────────────────────────────────────────────────────

class PRNG:
    """Fast PRNG implementing xoshiro256**."""

    MASK = (1 << 64) - 1

    def __init__(self, seed: int) -> None:
        self.s = [0, 0, 0, 0]
        z = seed & self.MASK
        for i in range(4):
            z = (z + 0x9E3779B97F4A7C15) & self.MASK
            z = ((z ^ (z >> 30)) * 0xBF58476D1CE4E5B9) & self.MASK
            z = ((z ^ (z >> 27)) * 0x94D049BB133111EB) & self.MASK
            self.s[i] = (z ^ (z >> 31)) & self.MASK

    @staticmethod
    def _rotl(x: int, k: int) -> int:
        mask = (1 << 64) - 1
        return ((x << k) | (x >> (64 - k))) & mask

    def next(self) -> int:
        s = self.s
        result = (self._rotl((s[1] * 5) & self.MASK, 7) * 9) & self.MASK
        t = (s[1] << 17) & self.MASK
        s[2] ^= s[0]
        s[3] ^= s[1]
        s[1] ^= s[2]
        s[0] ^= s[3]
        s[2] ^= t
        s[3] = self._rotl(s[3], 45)
        for i in range(4):
            s[i] &= self.MASK
        return result

    def next_range(self, n: int) -> int:
        if n <= 0:
            return 0
        return self.next() % n

    def random_bytes(self, n: int) -> bytes:
        return bytes(self.next() & 0xFF for _ in range(n))

    def pick(self, items: list):
        if not items:
            return None
        return items[self.next_range(len(items))]


# ─── Hash (FNV-1a 64-bit) ───────────────────────────────────────────────────

def fnv1a64(data: bytes) -> int:
    h = 0xCBF29CE484222325
    mask = (1 << 64) - 1
    for b in data:
        h ^= b
        h = (h * 0x100000001B3) & mask
    return h


# ─── Coverage Tracker ────────────────────────────────────────────────────────

class CoverageTracker:
    """Edge coverage tracker using a fixed-size bitmap."""

    def __init__(self, size: int = 65536) -> None:
        self._bitmap = bytearray(size)
        self._edges_found = 0

    def record_edge(self, from_: int, to: int) -> bool:
        idx = (from_ ^ to) % len(self._bitmap)
        if self._bitmap[idx] == 0:
            self._bitmap[idx] = 1
            self._edges_found += 1
            return True
        self._bitmap[idx] = min(self._bitmap[idx] + 1, 255)
        return False

    @property
    def edges_found(self) -> int:
        return self._edges_found

    @property
    def coverage_percent(self) -> float:
        return self._edges_found / len(self._bitmap) * 100

    def reset(self) -> None:
        for i in range(len(self._bitmap)):
            self._bitmap[i] = 0
        self._edges_found = 0


# ─── Mutator Engine ──────────────────────────────────────────────────────────

class MutatorStrategy:
    BIT_FLIP = "bit_flip"
    BYTE_FLIP = "byte_flip"
    ARITHMETIC = "arithmetic"
    INTERESTING = "interesting"
    HAVOC = "havoc"


class MutatorEngine:
    """Applies mutations to input bytes."""

    STRATEGIES = [
        MutatorStrategy.BIT_FLIP,
        MutatorStrategy.BYTE_FLIP,
        MutatorStrategy.ARITHMETIC,
        MutatorStrategy.INTERESTING,
        MutatorStrategy.HAVOC,
    ]

    INTERESTING_BYTES = [0, 1, 0x7F, 0x80, 0xFF]

    def __init__(self, prng: PRNG) -> None:
        self.prng = prng

    def mutate(self, data: bytes) -> bytes:
        if not data:
            return self.prng.random_bytes(self.prng.next_range(64) + 1)

        buf = bytearray(data)
        strategy = self.prng.pick(self.STRATEGIES)

        if strategy == MutatorStrategy.BIT_FLIP:
            pos = self.prng.next_range(len(buf))
            bit = self.prng.next_range(8)
            buf[pos] ^= 1 << bit

        elif strategy == MutatorStrategy.BYTE_FLIP:
            pos = self.prng.next_range(len(buf))
            buf[pos] ^= 0xFF

        elif strategy == MutatorStrategy.ARITHMETIC:
            pos = self.prng.next_range(len(buf))
            delta = self.prng.next_range(70) - 35
            buf[pos] = (buf[pos] + delta) & 0xFF

        elif strategy == MutatorStrategy.INTERESTING:
            pos = self.prng.next_range(len(buf))
            buf[pos] = self.prng.pick(self.INTERESTING_BYTES)

        elif strategy == MutatorStrategy.HAVOC:
            rounds = self.prng.next_range(8) + 1
            for _ in range(rounds):
                pos = self.prng.next_range(len(buf))
                buf[pos] = self.prng.next() & 0xFF

        return bytes(buf)


# ─── Corpus Store ────────────────────────────────────────────────────────────

@dataclass
class FuzzInput:
    data: bytes
    hash: int
    depth: int = 0
    energy: float = 1.0
    executions: int = 0
    created_at: float = field(default_factory=time.time)


class CorpusStore:
    """Manages the input corpus."""

    def __init__(self) -> None:
        self._entries: list[FuzzInput] = []
        self._hashes: set[int] = set()

    def add(self, inp: FuzzInput) -> bool:
        if inp.hash in self._hashes:
            return False
        self._hashes.add(inp.hash)
        self._entries.append(inp)
        return True

    def pick(self, prng: PRNG) -> Optional[FuzzInput]:
        return prng.pick(self._entries)

    @property
    def size(self) -> int:
        return len(self._entries)


# ─── Crash Analyzer ──────────────────────────────────────────────────────────

@dataclass
class CrashInfo:
    id: str
    input: FuzzInput
    error: str
    is_duplicate: bool
    discovered_at: float = field(default_factory=time.time)


class CrashAnalyzer:
    """Detects and deduplicates crashes."""

    def __init__(self) -> None:
        self._crashes: list[CrashInfo] = []
        self._hashes: set[int] = set()

    def record(self, inp: FuzzInput, error: str) -> bool:
        h = fnv1a64(error.encode())
        is_dup = h in self._hashes
        self._hashes.add(h)
        self._crashes.append(CrashInfo(
            id=f"crash-{h:016x}",
            input=inp,
            error=error,
            is_duplicate=is_dup,
        ))
        return not is_dup

    @property
    def unique_count(self) -> int:
        return len(self._hashes)


# ─── Fuzz Engine ─────────────────────────────────────────────────────────────

@dataclass
class FuzzStats:
    total_executions: int = 0
    corpus_size: int = 0
    unique_crashes: int = 0
    edges_found: int = 0
    coverage_percent: float = 0.0
    elapsed_ms: int = 0


@dataclass
class FuzzConfig:
    name: str = "unnamed"
    seed: int = 42
    max_executions: int = 0
    max_time_seconds: int = 0
    max_input_size: int = 4096
    target: Optional[Callable[[bytes], None]] = None


class FuzzEngine:
    """The main fuzzing engine."""

    def __init__(self, config: FuzzConfig) -> None:
        self.config = config
        self.prng = PRNG(config.seed)
        self.mutator = MutatorEngine(self.prng)
        self.corpus = CorpusStore()
        self.coverage = CoverageTracker(65536)
        self.crashes = CrashAnalyzer()
        self._stop = False

    def run(self) -> FuzzStats:
        start = time.monotonic()
        self._stop = False

        # Seed corpus
        if self.corpus.size == 0:
            seed_data = self.prng.random_bytes(self.prng.next_range(64) + 1)
            self.corpus.add(FuzzInput(data=seed_data, hash=fnv1a64(seed_data)))

        exec_count = 0

        while not self._stop:
            if self.config.max_executions > 0 and exec_count >= self.config.max_executions:
                break
            if self.config.max_time_seconds > 0:
                elapsed = time.monotonic() - start
                if elapsed > self.config.max_time_seconds:
                    break

            parent = self.corpus.pick(self.prng)
            if parent is None:
                break

            mutated = self.mutator.mutate(parent.data)
            h = fnv1a64(mutated)
            inp = FuzzInput(data=mutated, hash=h, depth=parent.depth + 1, energy=parent.energy)

            # Execute
            crashed = False
            error_msg = None
            if self.config.target:
                try:
                    self.config.target(mutated)
                except Exception as e:
                    crashed = True
                    error_msg = str(e)

            exec_count += 1

            # Coverage
            edge_hash = fnv1a64(mutated[:min(len(mutated), 8)])
            if self.coverage.record_edge(parent.hash, edge_hash):
                self.corpus.add(inp)

            # Crash
            if crashed and error_msg:
                self.crashes.record(inp, error_msg)

        elapsed_ms = int((time.monotonic() - start) * 1000)
        return FuzzStats(
            total_executions=exec_count,
            corpus_size=self.corpus.size,
            unique_crashes=self.crashes.unique_count,
            edges_found=self.coverage.edges_found,
            coverage_percent=self.coverage.coverage_percent,
            elapsed_ms=elapsed_ms,
        )

    def stop(self) -> None:
        self._stop = True
